### PR TITLE
Moves _getAuth and _getExternal functionality into the user class

### PR DESCRIPTION
**Pull request by jaycee**

### CODE REVIEW

**Comment by jaycee** [comment]
QA:

Spin up the gui against GIJoe and GISF.

Make sure the username is displayed properly in breadcrumbs. Make sure switching between models works. Make sure opening your profile page and other profile pages loads the correct data.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7365/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7366/

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -16,6 +16,8 @@ You should have received a copy of the GNU Affero General Public License along
 with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
 
+// TODO sort out get('auth') and get('users')
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: card/issue?

File: jujugui/static/gui/src/app/app.js [hunk 2]
```diff
@@ -1285,12 +1286,14 @@ YUI.add('juju-gui', function(Y) {
...
       }
       const getUserName = () => {
-        const credentials = this.user.controller;
-        return credentials ? credentials.user : undefined;
+        return this.user.username;
       };
       const loginToController = controllerAPI.loginWithMacaroon.bind(
         controllerAPI, this.bakery);
       const charmstore = this.get('charmstore');
+      const isLoggedIn = () => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: arrow fn's allow you to drop the `return` and curlies if you only have a single statement, up to you, just mentioning it.

```JavaScript
const isLoggedIn = () => this.controllerAPI.userIsAuthenticated;
```

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -1998,12 +2001,13 @@ YUI.add('juju-gui', function(Y) {
           // TODO frankban: here we should put a notification, but we can't as
           // this seems to be dispatched twice.
           console.log(msg);
+          // replace get auth with user.username
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: comment no longer needed?

File: jujugui/static/gui/src/app/app.js [hunk 4]
```diff
@@ -3019,78 +3023,7 @@ YUI.add('juju-gui', function(Y) {
...
       }
-    },
-
-    /**
-      A single point for accessing auth information that properly handles
-      situations where auth is set outside the GUI (i.e., embedded).
-
-      @method _getAuth
-     */
-    _getAuth: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: yay!

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js [hunk 5]
```diff
@@ -87,7 +87,7 @@ YUI.add('deployment-flow', function() {
         deploying: false,
         credential: this.props.credential,
         loadingTerms: false,
-        loggedIn: !!this.props.getAuth(),
+        loggedIn: !!this.props.isLoggedIn(),
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: `isLoggedIn` returns a boolean so the !! is not necessary.

before `getAuth` returned a non-boolean value so !! would type coerce

File: jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js [hunk 6]
```diff
@@ -138,9 +138,9 @@ YUI.add('header-breadcrumb', function() {
       @method _generateUserLink
     */
     _generateUserLink: function() {
-      const auth = this.props.authDetails;
-      if (auth && (auth.user || auth.loading)) {
-        return this._buildProfile(auth.loading ? '...' : auth.rootUserName);
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js]: Do we no longer need the ellipsis? I note no changes were made to the tests about this removal so either we didn't test this properly or we didn't use it....Just wondering which :)

File: jujugui/static/gui/src/app/user/test-user.js [hunk 7]
```diff
@@ -164,4 +164,28 @@ describe('user auth class', () => {
       assert.deepEqual(storage.getItem('test'), null);
     });
   });
+
+  describe('usernames', () =>{
```
Comment (hatched) [inline jujugui/static/gui/src/app/user/test-user.js]: thanks

File: jujugui/static/gui/src/app/user/user.js [hunk 8]
```diff
@@ -29,9 +29,37 @@ const User = class User {
     // We pass in these values to make test setup easier.
     this.sessionStorage = cfg.sessionStorage || sessionStorage;
     this.localStorage = cfg.localStorage || localStorage;
+    this._external = cfg.externalAuth || null;
   }
 
-  // TODO get username
+  get displayName() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/user/user.js]: So much easier to understand than the old method thx!

File: jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js (hunk 6 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js]: I think we can keep the ellipsis while waiting for the user to be available?

File: jujugui/static/gui/src/app/user/user.js (hunk 8 above)
Comment (frankban) [inline jujugui/static/gui/src/app/user/user.js]: Indeed! Please ad docstrings for these props.

**Comment by frankban** [comment]
QA:
when loading a profile page I get `failed prop type: Invalid prop `user` of type `string` supplied to `<<anonymous>>`, expected `object`. Other than that, QA is ok on both gisf and gijoe.

File: jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js (hunk 6 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/components/header-breadcrumb/header-breadcrumb.js]: The `auth` object there had no `loading` attribute, so the ellipsis was never shown. So I just removed it. At this point once you've hit the canvas you're logged into a controller, even if the model is still loading.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Oh, this is just leftover TODOs from my work I need to remove.

File: jujugui/static/gui/src/app/app.js (hunk 2 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Oh, cool.

File: jujugui/static/gui/src/app/app.js (hunk 3 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Correct. Removed.

File: jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js (hunk 5 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/components/deployment-flow/deployment-flow.js]: Ah, good catch.

**Comment by jaycee** [comment]
Thanks for the reviews!

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**